### ISSUE TITLE
WebTransport  with Certificate Hashes

### ISSUE THREAD

**Issue by lidel** [ipfs/in-web-browsers#190]
tl;dr
- WebTransport can act as a replacement for Secure WebSockets (`/wss`) that is easier to set up thanks to "certificate hashes" feature which might provide means to remove the need for getting TLS cert from some CA
- Limited to Chromium-based browsers now (Firefox wip but not ETA, Safari plans tbd)

## Status

- 2022-Q1: 
  - The WebTransport protocol is currently under development at the IETF. 
  - Chrome has implemented and shipped support for [draft-02](https://datatracker.ietf.org/doc/draft-ietf-webtrans-http3/) 
  - Chrome v100 [supports `serverCertificateHashes` option](https://chromestatus.com/feature/5690646332440576)
  - libp2p team is looking into "certificate hashes", internal evaluation ([libp2p notes](https://www.notion.so/pl-strflt/Enabling-WebTransport-across-the-libp2p-network-c6849c7252a1469a828a6c3e4b6abcad))
    - > The most exciting feature for us (other than the improved performance and hole punching success rates that QUIC gives us) is that the W3C added a browser API allowing browsers to establish connections to nodes with self-signed certificates, provided they know the hash of the certificate in advance: [`serverCertificateHashes`](https://www.w3.org/TR/webtransport/#dom-webtransportoptions-servercertificatehashes). Even more exciting, this is already [implemented in Chrome](https://chromestatus.com/feature/5690646332440576)  and will ship with version 100.

## Reading
- https://web.dev/webtransport/
- https://w3c.github.io/webtransport/
- https://github.com/w3c/webtransport/blob/main/explainer.md
- `serverCertificateHashes`  option 
  - https://chromestatus.com/feature/5690646332440576
  - https://w3c.github.io/webtransport/#dom-webtransportoptions-servercertificatehashes
  - https://groups.google.com/a/chromium.org/g/blink-dev/c/m0v9XiwKA4M
- https://www.notion.so/pl-strflt/Enabling-WebTransport-across-the-libp2p-network-c6849c7252a1469a828a6c3e4b6abcad
- Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1692754

**Comment by lidel** [comment]
WebTransport for libp2p is discussed in https://github.com/libp2p/specs/pull/404 :rocket:

**Comment by lidel** [comment]
We now have `/webtransport` transport protocol in libp2p.
- Specification https://github.com/libp2p/specs/blob/master/webtransport/README.md
-  It is enabled by default in [Kubo](https://github.com/ipfs/kubo) which listens on `*/udp/4001/quic-v1/webtransport/certhash/{certhash}` by default.

**Comment by lidel** [comment]
Firefox support for `/webtransport/certhash` is tracked in https://bugzilla.mozilla.org/show_bug.cgi?id=1806693

**Comment by lidel** [comment]
WIP patch with Firefox support for `/webtransport/certhash`: https://phabricator.services.mozilla.com/D195966

**Comment by lidel** [comment]
- https://github.com/w3c/webtransport/issues/589

**Comment by 2color** [comment]
An ongoing discussion to remove `serverCertificateHashes`: https://github.com/w3c/webtransport/issues/623
